src/runtime/typemanager.cs

```diff
@@ -817,4 +963,183 @@ internal static void CopySlot(IntPtr from, IntPtr to, int offset)
             Marshal.WriteIntPtr(to, offset, fp);
         }
     }
+
+    class SlotsHolder
+    {
+        public const string HolderKeyName = "_slots_holder";
+        public delegate void Resetor(IntPtr type, int offset);
+
+        private GCHandle _handle;
+        private Interop.DestructorFunc _destructor;
+        private IntPtr _capsule;
+        private IntPtr _type;
+        private Dictionary<int, ThunkInfo> _slots = new Dictionary<int, ThunkInfo>();
+        private List<Delegate> _keepalive = new List<Delegate>();
+        private Dictionary<int, Resetor> _customRestors = new Dictionary<int, Resetor>();
+        private List<Action> _deallocators = new List<Action>();
+        private bool _alredyReset = false;
+
+        /// <summary>
+        /// Create slots holder for holding the delegate of slots and be able  to reset them.
+        /// </summary>
+        /// <param name="type">Steals a reference to target type</param>
+        public SlotsHolder(IntPtr type)
+        {
+            _type = type;
+        }
+
+        public void Add(int offset, ThunkInfo thunk)
+        {
+            _slots.Add(offset, thunk);
+        }
+
+        public void Add(int offset, Resetor resetor)
+        {
+            _customRestors[offset] = resetor;
+        }
+
+        public void AddDealloctor(Action deallocate)
+        {
+            _deallocators.Add(deallocate);
+        }
+
+        public void KeeapAlive(Delegate d)
+        {
+            _keepalive.Add(d);
+        }
+
+        public IntPtr ToCapsule()
+        {
+            if (_capsule != IntPtr.Zero)
+            {
+                Runtime.XIncref(_capsule);
+                return _capsule;
+            }
+            _handle = GCHandle.Alloc(this);
+            _destructor = OnDestruct;
+            var fp = Marshal.GetFunctionPointerForDelegate(_destructor);
+            _capsule = Runtime.PyCapsule_New((IntPtr)_handle, null, fp);
+            return _capsule;
+        }
+
+        public static void ReleaseTypeSlots(IntPtr type)
+        {
+            IntPtr capsule = Runtime.PyObject_GetAttrString(type, HolderKeyName);
+            if (capsule == IntPtr.Zero)
+            {
+                return;
+            }
+            var self = RecoverFromCapsule(capsule);
+            self.ResetSlots();
+            Runtime.XDecref(capsule);
+
+            IntPtr tp_dict = Marshal.ReadIntPtr(type, TypeOffset.tp_dict);
+            if (Runtime.PyDict_DelItemString(tp_dict, HolderKeyName) != 0)
+            {
+                throw new PythonException();
+            }
+        }
+
+        private void ResetSlots()
+        {
+            if (_alredyReset)
+            {
+                return;
+            }
+            _alredyReset = true;
+            IntPtr tp_name = Marshal.ReadIntPtr(_type, TypeOffset.tp_name);
+            string typeName = Marshal.PtrToStringAnsi(tp_name);
```

Comment:
Judging by the code around typemanager.cs:430, the value of `tp_name` can well be a Unicode string. In Python3 they clearly use UTF8.

Reply:
This variable isn't used. We should remove it (and tp_name), and also remove the commented-out use at line 1055.

Reply:
Just some debug code I left...